File: src/routes/payments.ts
import { Router, Request, Response } from 'express';
import { db } from '@/services/database';
import { x402Service } from '@/services/x402';
import { requireAuth, requireActiveUser, getCurrentUser } from '@/middleware/auth';
import { validate, schemas } from '@/utils/validation';
import { ApiResponse, Payment, PaymentCreate, X402PaymentRequest, X402PaymentResponse } from '@/types';
import { logger } from '@/utils/logger';

const router = Router();

/**
 * POST /api/v1/payments
 * Create a new payment
 */
router.post('/', requireActiveUser, validate(schemas.paymentCreate), async (req: Request, res: Response) => {
  try {
    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const paymentData: PaymentCreate = {
      ...req.body,
      payerId: currentUser.id, // Override with current user ID
    };

    const payment = await db.payment.create(paymentData);

    const response: ApiResponse<Payment> = {
      success: true,
      data: payment,
      message: 'Payment created successfully',
    };

    res.status(201).json(response);
  } catch (error) {
    logger.error('Create payment failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to create payment',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/payments
 * Get all payments for current user
 */
router.get('/', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const payments = await db.payment.findByPayerId(currentUser.id);

    const response: ApiResponse<Payment[]> = {
      success: true,
      data: payments,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get payments failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get payments',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/payments/:id
 * Get payment by ID
 */
router.get('/:id', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const paymentId = parseInt(req.params.id);
    
    if (isNaN(paymentId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid payment ID',
      };
      return res.status(400).json(response);
    }

    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const payment = await db.payment.findById(paymentId);
    
    if (!payment) {
      const response: ApiResponse = {
        success: false,
        message: 'Payment not found',
      };
      return res.status(404).json(response);
    }

    // Check if user is authorized to view this payment
    if (payment.payerId !== currentUser.id) {
      const response: ApiResponse = {
        success: false,
        message: 'Unauthorized to view this payment',
      };
      return res.status(403).json(response);
    }

    const response: ApiResponse<Payment> = {
      success: true,
      data: payment,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get payment failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get payment',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * PUT /api/v1/payments/:id
 * Update payment
 */
router.put('/:id', requireActiveUser, validate(schemas.paymentUpdate), async (req: Request, res: Response) => {
  try {
    const paymentId = parseInt(req.params.id);
    
    if (isNaN(paymentId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid payment ID',
      };
      return res.status(400).json(response);
    }

    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    // Check if payment exists and user is authorized
    const existingPayment = await db.payment.findById(paymentId);
    
    if (!existingPayment) {
      const response: ApiResponse = {
        success: false,
        message: 'Payment not found',
      };
      return res.status(404).json(response);
    }

    if (existingPayment.payerId !== currentUser.id) {
      const response: ApiResponse = {
        success: false,
        message: 'Unauthorized to update this payment',
      };
      return res.status(403).json(response);
    }

    const updatedPayment = await db.payment.update(paymentId, req.body);
    
    if (!updatedPayment) {
      const response: ApiResponse = {
        success: false,
        message: 'Failed to update payment',
      };
      return res.status(500).json(response);
    }

    const response: ApiResponse<Payment> = {
      success: true,
      data: updatedPayment,
      message: 'Payment updated successfully',
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Update payment failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to update payment',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/payments/rent-agreement/:rentAgreementId
 * Get payments for a specific rent agreement
 */
router.get('/rent-agreement/:rentAgreementId', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const rentAgreementId = parseInt(req.params.rentAgreementId);
    
    if (isNaN(rentAgreementId)) {
      const response: ApiResponse = {
        success: false,
        message: 'Invalid rent agreement ID',
      };
      return res.status(400).json(response);
    }

    const currentUser = getCurrentUser(req);
    
    if (!currentUser) {
      const response: ApiResponse = {
        success: false,
        message: 'User not found',
      };
      return res.status(404).json(response);
    }

    const payments = await db.payment.findByRentAgreementId(rentAgreementId);

    // Filter payments to only show those belonging to the current user
    const userPayments = payments.filter(payment => payment.payerId === currentUser.id);

    const response: ApiResponse<Payment[]> = {
      success: true,
      data: userPayments,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get rent agreement payments failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get rent agreement payments',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * POST /api/v1/payments/x402/create
 * Create X402 payment
 */
router.post('/x402/create', requireActiveUser, validate(schemas.x402PaymentRequest), async (req: Request, res: Response) => {
  try {
    const paymentRequest: X402PaymentRequest = req.body;
    const x402Payment = await x402Service.createPayment(paymentRequest);

    const response: ApiResponse<X402PaymentResponse> = {
      success: true,
      data: x402Payment,
      message: 'X402 payment created successfully',
    };

    res.status(201).json(response);
  } catch (error) {
    logger.error('Create X402 payment failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to create X402 payment',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * GET /api/v1/payments/x402/:paymentId/status
 * Get X402 payment status
 */
router.get('/x402/:paymentId/status', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const { paymentId } = req.params;
    const status = await x402Service.getPaymentStatus(paymentId);

    const response: ApiResponse<typeof status> = {
      success: true,
      data: status,
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Get X402 payment status failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to get X402 payment status',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * POST /api/v1/payments/x402/callback
 * Handle X402 payment webhook
 */
router.post('/x402/callback', async (req: Request, res: Response) => {
  try {
    const webhookData = req.body;
    const processedWebhook = await x402Service.processWebhook(webhookData);

    // Update payment status in database if needed
    // This would typically involve finding the payment by reference and updating its status

    const response: ApiResponse<typeof processedWebhook> = {
      success: true,
      data: processedWebhook,
      message: 'Webhook processed successfully',
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Process X402 webhook failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to process webhook',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

/**
 * POST /api/v1/payments/x402/:paymentId/refund
 * Refund X402 payment
 */
router.post('/x402/:paymentId/refund', requireActiveUser, async (req: Request, res: Response) => {
  try {
    const { paymentId } = req.params;
    const { amount } = req.body;

    const refund = await x402Service.refundPayment(paymentId, amount);

    const response: ApiResponse<typeof refund> = {
      success: true,
      data: refund,
      message: 'Payment refunded successfully',
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error('Refund X402 payment failed:', error);
    
    const response: ApiResponse = {
      success: false,
      message: 'Failed to refund payment',
      error: error instanceof Error ? error.message : 'Unknown error',
    };

    res.status(500).json(response);
  }
});

export default router;

File: app/routers/payments.py
from fastapi import APIRouter, Depends, HTTPException, status, Request
from sqlalchemy.orm import Session
from typing import List, Dict, Any

from app.database import get_db
from app import crud, schemas, auth, models
from app.x402_service import x402_service

router = APIRouter()

@router.post("/", response_model=schemas.PaymentResponse)
def create_payment(
    payment: schemas.PaymentCreate,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Create a new payment"""
    # Check if rent agreement exists
    agreement = crud.get_rent_agreement(db, payment.rent_agreement_id)
    if not agreement:
        raise HTTPException(status_code=404, detail="Rent agreement not found")
    
    # Check permissions - either tenant or property owner can create payments
    property = crud.get_property(db, agreement.property_id)
    if current_user.id != agreement.tenant_id and current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    # Check if payer exists
    payer = crud.get_user(db, payment.payer_id)
    if not payer:
        raise HTTPException(status_code=404, detail="Payer not found")
    
    return crud.create_payment(db=db, payment=payment)

@router.post("/{payment_id}/pay-with-x402", response_model=schemas.X402PaymentResponse)
async def pay_with_x402(
    payment_id: int,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Initiate X402 payment for a payment record"""
    # Get payment
    payment = crud.get_payment(db, payment_id)
    if not payment:
        raise HTTPException(status_code=404, detail="Payment not found")
    
    # Check if current user is the payer
    if current_user.id != payment.payer_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only the payer can initiate payment"
        )
    
    # Check if payment is still pending
    if payment.status != models.PaymentStatus.PENDING:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Payment is not in pending status"
        )
    
    try:
        # Create X402 payment request
        x402_request = schemas.X402PaymentRequest(
            amount=payment.amount,
            currency=payment.currency,
            description=f"Rent payment - {payment.description}",
            payment_reference=payment.payment_reference
        )
        
        # Initiate X402 payment
        x402_response = await x402_service.create_payment(x402_request)
        
        # Update payment with X402 details
        crud.update_payment_x402_details(
            db, 
            payment_id, 
            x402_response.payment_id
        )
        
        # Update payment method
        payment_update = schemas.PaymentUpdate(payment_method="x402")
        crud.update_payment(db, payment_id, payment_update)
        
        return x402_response
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"X402 payment initiation failed: {str(e)}"
        )

@router.post("/x402/callback")
async def x402_webhook_callback(request: Request, db: Session = Depends(get_db)):
    """Handle X402 webhook callbacks"""
    try:
        webhook_data = await request.json()
        
        # Process webhook
        processed_data = await x402_service.process_webhook(webhook_data)
        
        # Find payment by X402 payment ID
        payment = db.query(models.Payment).filter(
            models.Payment.x402_payment_id == processed_data["payment_id"]
        ).first()
        
        if not payment:
            raise HTTPException(status_code=404, detail="Payment not found")
        
        # Update payment status based on webhook
        status_map = {
            "completed": models.PaymentStatus.COMPLETED,
            "failed": models.PaymentStatus.FAILED,
            "refunded": models.PaymentStatus.REFUNDED
        }
        
        new_status = status_map.get(processed_data["status"], models.PaymentStatus.PENDING)
        
        payment_update = schemas.PaymentUpdate(
            status=new_status
        )
        
        # Update transaction hash if provided
        if processed_data.get("transaction_hash"):
            crud.update_payment_x402_details(
                db,
                payment.id,
                payment.x402_payment_id,
                processed_data["transaction_hash"]
            )
        
        crud.update_payment(db, payment.id, payment_update)
        
        return {"status": "success", "message": "Webhook processed successfully"}
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Webhook processing failed: {str(e)}"
        )

@router.get("/my-payments", response_model=List[schemas.PaymentResponse])
def get_my_payments(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get payments made by current user"""
    return crud.get_payments_by_payer(db=db, payer_id=current_user.id, skip=skip, limit=limit)

@router.get("/agreement/{agreement_id}", response_model=List[schemas.PaymentResponse])
def get_payments_by_agreement(
    agreement_id: int,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get payments for a specific rent agreement"""
    # Check if agreement exists
    agreement = crud.get_rent_agreement(db, agreement_id)
    if not agreement:
        raise HTTPException(status_code=404, detail="Rent agreement not found")
    
    # Check permissions
    property = crud.get_property(db, agreement.property_id)
    if current_user.id != agreement.tenant_id and current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    return crud.get_payments_by_agreement(db=db, agreement_id=agreement_id, skip=skip, limit=limit)

@router.get("/{payment_id}", response_model=schemas.PaymentResponse)
def get_payment(
    payment_id: int,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get payment by ID"""
    payment = crud.get_payment(db, payment_id=payment_id)
    if payment is None:
        raise HTTPException(status_code=404, detail="Payment not found")
    
    # Check permissions - payer or property owner can view
    agreement = crud.get_rent_agreement(db, payment.rent_agreement_id)
    property = crud.get_property(db, agreement.property_id)
    
    if current_user.id != payment.payer_id and current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    return payment

@router.get("/", response_model=List[schemas.PaymentResponse])
def get_payments(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get list of all payments"""
    payments = crud.get_payments(db, skip=skip, limit=limit)
    return payments

@router.get("/status/pending", response_model=List[schemas.PaymentResponse])
def get_pending_payments(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get pending payments"""
    return crud.get_pending_payments(db=db, skip=skip, limit=limit)

@router.put("/{payment_id}", response_model=schemas.PaymentResponse)
def update_payment(
    payment_id: int,
    payment_update: schemas.PaymentUpdate,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Update payment information"""
    # Check if payment exists
    payment = crud.get_payment(db, payment_id=payment_id)
    if payment is None:
        raise HTTPException(status_code=404, detail="Payment not found")
    
    # Check permissions - only property owner can update payments
    agreement = crud.get_rent_agreement(db, payment.rent_agreement_id)
    property = crud.get_property(db, agreement.property_id)
    
    if current_user.id != property.owner_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only property owner can update payments"
        )
    
    updated_payment = crud.update_payment(db, payment_id=payment_id, payment_update=payment_update)
    return updated_payment

@router.get("/{payment_id}/x402-status")
async def get_x402_payment_status(
    payment_id: int,
    db: Session = Depends(get_db),
    current_user: schemas.UserResponse = Depends(auth.get_current_active_user)
):
    """Get X402 payment status"""
    payment = crud.get_payment(db, payment_id)
    if not payment:
        raise HTTPException(status_code=404, detail="Payment not found")
    
    # Check permissions
    if current_user.id != payment.payer_id:
        agreement = crud.get_rent_agreement(db, payment.rent_agreement_id)
        property = crud.get_property(db, agreement.property_id)
        if current_user.id != property.owner_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not enough permissions"
            )
    
    if not payment.x402_payment_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Payment was not made through X402"
        )
    
    try:
        x402_status = await x402_service.get_payment_status(payment.x402_payment_id)
        return x402_status
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get X402 status: {str(e)}"
        )
